File: applied/6_born_in_the_usa/answer/high.c
// 入力された文字列について、無限にしりとりが成立するか判定しよう

#include <stdio.h>
int main(void){
    int i;
    int start[26] = {0};
    int end[26] = {0};
    char input[5];
    
    for (scanf("%s", input); input[0] != '0'; scanf("%s", input)) {
        start[input[0] - 0x61]++;
        end[input[4] - 0x61]++;
    }
    for (i=0; i<26; i++) {
        if (start[i] != end[i]) break;
    }
    if (i == 26) {
        printf("yes\n");
    } else {
        printf("no\n");
    }
    
    return 0;
}

File: applied/6_born_in_the_usa/answer/high.java
// 入力された文字列について、無限にしりとりが成立するか判定しよう

import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        Scanner scan = new Scanner(System.in);
        String input;
        int[] start = new int[26];
        int[] end = new int[26];
        
        for (input = scan.next(); !input.equals("0"); input = scan.next()) {
            start[input.charAt(0) - 'a']++;
            end[input.charAt(4) - 'a']++;
        }
        
        int i;
        for (i=0; i<26; i++) {
            if (start[i] != end[i]) break;
        }
    
        if (i == 26) {
            System.out.printf("yes\n");
        } else {
            System.out.printf("no\n");
        }
    }
}
